{% extends 'global/base.html' %}
{% load static %}

{% block styles %}
    <!-- Site-specific css/js -->
    <link rel="stylesheet" type="text/css" href="{% static 'global/css/annotype_color.css' %}">
    <link rel="stylesheet" type="text/css" href="{% static 'global/css/taxid_color.css' %}" id="taxid-color-stylesheet"/>
    <link rel="stylesheet" type="text/css" href="{% static 'global/css/tag_color.css' %}"/>
    <link rel="stylesheet" type="text/css" href="{% static 'gene-trait-matching/css/gene-trait-matching.css' %}"/>
    <script src="{% static 'global/js/query-groups.js' %}"></script>

    <!-- autocomplete -->
    <link href="{% static 'global/css/jquery.tag-editor.css' %}" rel="stylesheet"/>
    <script src="{% static 'global/js/jquery.caret.min.js' %}"></script>
    <script src="{% static 'global/js/jquery.tag-editor.min.js' %}"></script>

    <!-- DataTables -->
    <script type="text/javascript" src="{% static 'global/js/datatables.min.js' %}"></script>

{% endblock %}

{% block sidebar %}

    <div class="sidebar-subcontent form-check form-check-inline">
        <label class="form-check-label" for="colorize-tax-checkbox"><a>Colorize taxonomy?</a></label>
        <input class="form-check-input" type="checkbox" id="colorize-tax-checkbox" onclick="toggleColorizeTax()" checked>
    </div>

    <div class="sidebar-subcontent">
        <div id="buttons" class="sidebar-subcontent">
            <a>Export table: </a>
        </div>
    </div>


    <hr>


    <div class="sidebar-subcontent form-check form-check-inline">
        <label for="gene-trait-matching-anno-type"><a>Which category of annotations to use?</a></label>
        <select class="form-control" id="gene-trait-matching-anno-type" data-selected="{{ default_anno_type }}">
            {% for at in anno_types %}
                <option value="{{ at.anno_type }}" {% if at.anno_type == default_anno_type %} selected {% endif %}>{{ at.name }} ({{ at.anno_type }})
                </option>{% endfor %}
        </select>
    </div>

    <div class="sidebar-subcontent form-check form-check-inline">
        <label for="gene-trait-matching-multiple-testing-method"><a>Multiple testing correction </a><a style="font-size:small"
                                                                                                       href="https://www.statsmodels.org/dev/generated/statsmodels.stats.multitest.multipletests.html">(click
            here for more information)</a></label>
        <select class="form-control" id="gene-trait-matching-multiple-testing-method" data-selected="{{ anno_type }}">
            {% for mtm_val, mtm_descr in multiple_testing_methods.items %}
                <option value="{{ mtm_val }}" {% if mtm_val == default_multiple_testing_method %} selected {% endif %}>{{ mtm_descr }}</option>
            {% endfor %}
        </select>
    </div>

    <div class="sidebar-subcontent form-check form-check-inline">
        <label class="form-check-label" for="form-control-range"><a>Alpha / FWER (family-wise error rate)</a></label>
        <input type="range" class="form-control-range" id="form-control-range" min="0.01" max="0.99" value="{{ default_alpha }}" step="0.01">
        <span>alpha = <b id="range-span">{{ default_alpha }}</b></span>
    </div>

    <hr>

    <div class="sidebar-subcontent form-check form-check-inline">
        <button type="button" class="btn btn-primary" onclick="$('#gene-trait-matching-submit-button').click()">Submit</button>
    </div>


{% endblock %}

{% block body %}

    <section>

        <div class="container">
            <form>
                {% csrf_token %}
                <div id="query-genomes"></div>

                <button id="gene-trait-matching-submit-button" type="button" class="btn btn-primary" onclick="submitFunction()">Submit</button>
            </form>

            {% if magic_query_manager_g1 and magic_query_manager_g2 %}
                <hr class="featurette-divider">

                <button type="button" class="btn btn-light" data-toggle="collapse" href="#real-div" aria-expanded="false" aria-controls="real-div"
                >Show query
                </button>

                <form id="real-div" class="collapse">
                    <div class="form-group">
                        <label for="real-genomes-1">Group 1:</label>
                        <div class="read-only-div" id="real-genomes-1">
                            {% for genome in magic_query_manager_g1.all_genomes %}{{ genome.html|safe }}{% endfor %}
                        </div>
                    </div>

                    <div class="form-group">
                        <label for="real-genomes-2">Group 2:</label>
                        <div class="read-only-div" id="real-genomes-2">
                            {% for genome in magic_query_manager_g2.all_genomes %}{{ genome.html|safe }}{% endfor %}
                        </div>
                    </div>

                    <div class="form-group">
                        <label>Parameters:</label>
                        <div class="read-only-div" id="real-anno-type">
                            <a>Annotation type: {{ default_anno_type }}</a>
                            <a>Multiple testing method: {{ default_multiple_testing_method }}</a>
                            <a>Alpha: {{ default_alpha }}</a>
                        </div>
                    </div>

                </form>
                </div>
            {% endif %}

        </div>

    </section>

    <section style="padding-top: 1rem">
        <div id="annotation-matrix-container">
            <div style="display:flex; justify-content: center;">
                <div id="gene-loci-spinner" class="spinner-border text-dark" role="status" hidden>
                    <span class="sr-only">Loading...</span>
                </div>
            </div>
        </div>
    </section>

    <section id="gene-trait-matching-container" style="padding-top: 1rem">
        <div class="container">
            <table id="gene-trait-matching-table" class="table table-striped table-bordered" style="width:100%"></table>
        </div>
    </section>


{% endblock %}

{% block end %}
    <script>
        "use strict"

        const genomesG1 = [{% if magic_query_manager_g1 %} '{{ magic_query_manager_g1.queries|join:"', '" }}' {% else %}{% endif %}]
        const genomesG2 = [{% if magic_query_manager_g2 %} '{{ magic_query_manager_g2.queries|join:"', '" }}' {% else %}{% endif %}]

        const queryGenomes = $('#query-genomes')

        $(document).ready(function () {
            addGenomesGroup(queryGenomes, genomesG1, false)
            addGenomesGroup(queryGenomes, genomesG2, false)

            $('.ogb-tag.genome')
                .ogbTooltip()
                .click(function () {
                    showGenomeClickMenu(event)
                })

        })

        $(document).ready(function () {
            {% if success %}
                loadTable(genomesG1, genomesG2, {{default_alpha}}, "{{default_anno_type}}", "{{default_multiple_testing_method}}")
            {% endif %}
        })

        function submitFunction() {
            const getQueries = queryGenomes.find('.get-genomes')

            const genomes_g1 = extractQuery(getQueries[0])
            const genomes_g2 = extractQuery(getQueries[1])

            const alpha = parseFloat($('#range-span').text())
            const anno_type = $('#gene-trait-matching-anno-type').val()
            const multiple_testing_method = $('#gene-trait-matching-multiple-testing-method').val()

            $.when(validateGenomes(genomes_g1), validateGenomes(genomes_g2)).done(function (genomes_g1_valid, genomes_g2_valid) {
                genomes_g1_valid = genomes_g1_valid[2].responseJSON['success']
                genomes_g2_valid = genomes_g2_valid[2].responseJSON['success']

                if (genomes_g1_valid) {
                    $('.get-genomes').first().removeClass('is-invalid')
                } else {
                    $('.get-genomes').first().addClass('is-invalid')
                }

                if (genomes_g2_valid) {
                    $('.get-genomes').last().removeClass('is-invalid')
                } else {
                    $('.get-genomes').last().addClass('is-invalid')
                }

                if (genomes_g1_valid && genomes_g2_valid) {
                    let data = {
                        'alpha': alpha.toString(),
                        'anno_type': anno_type,
                        'multiple_testing_method': multiple_testing_method
                    }
                    if (genomes_g1.length !== 0) {
                        data['g1'] = genomes_g1
                    }
                    if (genomes_g2.length !== 0) {
                        data['g2'] = genomes_g2
                    }

                    goToPageWithData('/gene-trait-matching', data)
                }
            })
        }

        function activateTable(settings) {
            const table = $('#gene-trait-matching-table').DataTable()

            // click on annotations
            $('.ogb-tag.annotation').each(function (index) {
                // only add click event if there is none yet
                if (jQuery._data(this, 'events') === undefined) {
                    $(this)
                        .ogbTooltip()
                        .on('click', function () {
                            table.rows(this.parentElement.parentElement).select()  // select this row
                            const data = table.rows({selected: true}).data()
                            const selectedAnnotations = Array.from({length: data.length}, (_, i) => data[i][0])
                            showAnnotationClickMenu(event, 'auto', selectedAnnotations, {'Group 1': '#real-genomes-1', 'Group 2': '#real-genomes-2'})
                        })
                }
            })
        }

        /**
         * Enable scientific notation sorting
         */
        jQuery.extend(jQuery.fn.dataTableExt.oSort, {
            "scientific-pre": function (a) {
                return parseFloat(a);
            },

            "scientific-asc": function (a, b) {
                return ((a < b) ? -1 : ((a > b) ? 1 : 0));
            },

            "scientific-desc": function (a, b) {
                return ((a < b) ? 1 : ((a > b) ? -1 : 0));
            }
        });

        function drawTable(data) {
            const table = $('#gene-trait-matching-table').DataTable({
                data: data['dataset'],
                columns: data['columns'],
                select: true, multi: true,
                lengthMenu: [[-1, 10, 25, 50, 100], ['All', 10, 25, 50, 100]],
                pageLength: 25,
                order: [[7, 'asc']],  // default sorting: pvalue
                'drawCallback': activateTable,
                language: {
                    'emptyTable': 'No significant differences found. More genomes may be required.'
                },
                'columnDefs': [
                    // {0: '_annotation', 1: 'Description', 2: 'Annotation', 3: 'Group 1 [%]', 4: 'Group 2 [%]', 5: 'Group 1', 6: 'Group 2', 7: 'pv', 8: 'qv', 9: 'reject'}
                    {'targets': [0], 'visible': false, 'searchable': false}, // _annotation
                    {'targets': [1], 'visible': false}, // Description
                    {'targets': [3, 3, 4, 5, 6, 9], 'searchable': false}, // numbers, reject
                    {'targets': [7, 8], type: 'scientific', 'searchable': false}, // pvalue, qvalue
                ]
            })

            const buttons = new $.fn.dataTable.Buttons(table, {
                buttons: [{
                    extend: 'csv',
                    exportOptions: {
                        columns: ':visible'  // download visibe columns only
                    }
                }]
            }).container().appendTo($('#buttons'))

        }


        async function loadTable(genomes_g1, genomes_g2, alpha, anno_type, multiple_testing_method) {
            if (genomes_g1.length === 0 && genomes_g2.length === 0) {
                console.log('no annotations or no genomes!')
                return
            }

            $('#gene-loci-spinner').attr('hidden', false)

            $.post('{% url 'website:gene-trait-matching-table' %}', {
                'g1[]': genomes_g1,
                'g2[]': genomes_g2,
                'alpha': alpha,
                'anno_type': anno_type,
                'multiple_testing_method': multiple_testing_method
            }, 'json')
                .done(function (data, textStatus, jqXHR) {
                    drawTable(data)
                    activateTable()
                })
                .fail(function (data, textStatus, jqXHR) {
                    if (data?.status === 409) {
                        console.log('empty df', data, textStatus, jqXHR)
                        alertModal('info', 'Info', data?.responseJSON?.message)
                    } else {
                        console.log('fail', data, textStatus, jqXHR)
                        alertModal('danger', 'Error', data?.responseJSON?.message || `(no message) ${data?.status} ${data?.statusText}`)
                    }
                })
                .always(function () {
                    $('#gene-loci-spinner').attr('hidden', true)
                })
        }

        $(document).ready(function () {
            $('#form-control-range').on('change', function (e) {
                const span = $('#form-control-range').val()
                $('#range-span').text(span)
            })
        })

    </script>

{% endblock %}
