## Welcome to OpenGenomeBrowser 
 
**Demo instance:** [opengenomebrowser.bioinformatics.unibe.ch](https://opengenomebrowser.bioinformatics.unibe.ch/)

To talk to the developer informally, join the OpenGenomeBrowser [Discord channel](https://discord.gg/mDm4fqf).

<hr>

_OpenGenomeBrowser is a dynamic and scalable web platform for comparative genomics._

*   OpenGenomeBrowser automates simple bioinformatics steps:
    - Filter and search using metadata
    - Dendrograms (Taxid-based, ANI-based and OrthoFinder-based)
    - Searching for annotations in genomes
    - Visualizing genomes on pathway maps
    - Visualisation of gene loci
    - BLAST
    - Access raw data
*   OpenGenomeBrowser has few prerequisites:
    - Genome files must be stored in certain folder structure
    - One metadata file for each genome and organism
*   OpenGenomeBrowser is flexible:
    - [OrthoFinder](https://github.com/davidemms/OrthoFinder) is recommended, but not required (allows core genome dendrogram calculations)
    - Computationally intensive processes can be outsourced to a cluster

<div align="center">
<img src="https://raw.githubusercontent.com/opengenomebrowser/opengenomebrowser/master/website/static/global/customicons/ogb-full.svg"  width="500px">
</div>


### Key concepts

*   _Organism_: biological entity
*   _Genome_: sequencing variant, belongs to an organism
*   _Representative_: best genome of organism


## Tutorial

Play around with the demo instance of [OpenGenomeBrowser](opengenomebrowser.bioinformatics.unibe.ch/) and read the [tutorial](/tutorial.md)!


## Installation

Click [here](/installation.md) to read the instructions.


## Get help

If you find a bug that has not already been reported, [submit an issuehere on GitHub](https://github.com/opengenomebrowser/opengenomebrowser/issues).

For informal chatting, join the OpenGenomeBrowser [Discord channel](https://discord.gg/mDm4fqf).
